File: tests/test_eligible_diagnostics.py
import ast,csv,json
from pathlib import Path
import sys,tempfile,unittest,subprocess
sys.path.insert(0,str(Path(__file__).resolve().parents[1]/'scripts'))
import launch_eligible_diagnostics as m
ROOT=Path(__file__).resolve().parents[1]
class EligibleTests(unittest.TestCase):
 def test_plan_nine_no_fits(self):
  with tempfile.TemporaryDirectory() as tmp:
   dest=Path(tmp)/'plan';p=m.prepare(ROOT,dest,'clean',False)
   self.assertEqual(len(p['tasks']),9);self.assertFalse(p['models_fitted'])
   for n in ('run.sh','collect.sh'):subprocess.check_call(['bash','-n',str(dest/n)])
   with self.assertRaisesRegex(ValueError,'Unverified'):m.verify(dest,p,m.sha(dest/'plan.json'))
 def test_reconciled_reports_and_tampering(self):
  with tempfile.TemporaryDirectory() as tmp:
   work=Path(tmp)/'CRYPTOSPORIDIUM';out=work/'result';out.mkdir(parents=True)
   def write(n,rr):
    with (out/n).open('w',newline='') as f:w=csv.DictWriter(f,fieldnames=list(rr[0]));w.writeheader();w.writerows(rr)
   annual=[dict(state=s,year=y,eligible_records=2,population=1000) for s in ('CA','CO','CT','GA','MD','MN','NM','NY','OR','TN') for y in range(2004,2018)]
   support=[dict(r,cx_classified=0,cidt_classified=0,parasitic_classified=2,classification_denominator=0,cidt_classification_share='') for r in annual]
   categories=[dict(state=r['state'],year=r['year'],category=c,records=2 if c=='PARASITIC' else 0) for r in annual for c in ('CX+','CIDT+','PARASITIC')]
   write('annual.csv',annual);write('support.csv',support);write('categories.csv',categories)
   write('readiness.csv',[dict(pathogen='CRYPTOSPORIDIUM',end_year=2017,counts_reconciled='TRUE',model_fitted='FALSE',coverage_certified='FALSE',selected_records=280)])
   write('flow.csv',[dict(records=n) for n in (282,2,280)]);write('input_checksums.csv',[dict(file='synthetic',md5='test')]);(out/'status.txt').write_text('ELIGIBLE_DIAGNOSTICS_COMPLETE')
   self.assertTrue(m.validate_result(work));support[0]['cidt_classification_share']='0';write('support.csv',support)
   with self.assertRaisesRegex(ValueError,'missing'):m.validate_result(work)
 def test_partial_collection(self):
  with tempfile.TemporaryDirectory() as tmp:
   dest=Path(tmp)/'plan';m.prepare(ROOT,dest,'clean',False)
   self.assertEqual(m.collect(dest,m.sha(dest/'plan.json')),1)
   self.assertTrue(Path(str(dest)+'.tar.gz').exists())
 def test_python36(self):ast.parse((ROOT/'scripts/launch_eligible_diagnostics.py').read_text(),feature_version=(3,6))
if __name__=='__main__':unittest.main()

File: tests/test_eligible_diagnostics.R
source('scripts/county_matching.R');source('scripts/audit_eligible_diagnostics.R')
p<-data.frame(state=c('AA','AA','AA'),fips=c('01001','01001','01001'),year=2015:2017,population=1000,count=c(2,1,0))
d<-data.frame(state='AA',fips='01001',year=c(2015,2015,2016),county='TEST',pathogen='LISTERIA',travelint='NO',cxcidt=c('CX+','CIDT+','CX+'),cste='YES')
x<-eligible_category_tables(d,p,'LISTERIA');stopifnot(sum(x$categories$records)==3,x$support$cidt_classification_share[1]==.5,is.na(x$support$cidt_classification_share[3]))
bad<-d;bad$cste[1]<-'NO';stopifnot(inherits(try(eligible_category_tables(bad,p,'LISTERIA'),silent=TRUE),'try-error'))
bad<-d;bad$fips[1]<-'01003';stopifnot(inherits(try(eligible_category_tables(bad,p,'LISTERIA'),silent=TRUE),'try-error'))
d$pathogen<-'CRYPTOSPORIDIUM';d$cxcidt<-'PARASITIC';extra<-d[1,];extra$year<-2018
x<-eligible_category_tables(rbind(d,extra),p,'CRYPTOSPORIDIUM');stopifnot(sum(x$categories$records)==3,all(x$support$classification_denominator==0),all(is.na(x$support$cidt_classification_share)))
p$count[1]<-3;stopifnot(inherits(try(eligible_category_tables(d,p,'CRYPTOSPORIDIUM'),silent=TRUE),'try-error'))
cat('Eligible category reconciliation, CSTE, Crypto cutoff and missing-share tests PASS\n')
# Exercise actual CSV ingestion and the production report writer on a synthetic full domain.
work<-tempfile();dir.create(work);audit<-file.path(work,'audit');dir.create(audit);dir.create(file.path(audit,'reports'))
panel<-expand.grid(fips=sprintf('%05d',1001:1486),year=2004:2017,stringsAsFactors=FALSE);panel$state<-'AA';panel$population<-10000;panel$count<-1L
validate_panel<-function(...)list(data=panel)
clean<-panel[c('fips','state','year')];clean$county<-'TEST';clean$pathogen<-'CRYPTOSPORIDIUM';clean$travelint<-'NO';clean$cxcidt<-'PARASITIC'
cp<-file.path(work,'clean.csv');write.csv(clean,cp,row.names=FALSE);saveRDS(panel,file.path(audit,'county_panel_INTERNAL.rds'))
write.csv(data.frame(file=cp,md5=unname(tools::md5sum(cp))),file.path(audit,'reports/input_checksums.csv'),row.names=FALSE)
out<-file.path(work,'result');run_audit(cp,audit,out,'CRYPTOSPORIDIUM')
stopifnot(readLines(file.path(out,'status.txt'))=='ELIGIBLE_DIAGNOSTICS_COMPLETE',sum(read.csv(file.path(out,'categories.csv'))$records)==6804)
cat('Actual eligible CSV/report integration PASS\n')
